app/src/main/java/de/triology/cas/services/RegistryLocal.java

```diff
@@ -217,31 +222,73 @@ public URI getCasLogoutUri(String doguname) throws GetCasLogoutUriException {
 
     @Override
     public void addDoguChangeListener(DoguChangeListener doguChangeListener) {
-        Thread t1 = new Thread(() -> {
-            var path = Paths.get(LOCAL_CONFIG_DIR);
-            try(WatchService watchService = fileSystem.newWatchService()) {
-                path.register(watchService, StandardWatchEventKinds.ENTRY_CREATE, StandardWatchEventKinds.ENTRY_MODIFY);
-                var previousServiceAccounts = readServiceAccounts();
-                while (true) {
-                    LOGGER.info("wait for changes under {}", LOCAL_CONFIG_DIR);
-                    watchService.take();
-                    var currentServiceAccounts = readServiceAccounts();
-                    if (!previousServiceAccounts.deepEquals(currentServiceAccounts)) {
-                        doguChangeListener.onChange();
-                        previousServiceAccounts = currentServiceAccounts;
+        Thread t1 = new Thread(() -> doguChangeListenerThreadHandler(doguChangeListener));
+        t1.start();
+    }
+
+    private void doguChangeListenerThreadHandler(DoguChangeListener doguChangeListener) {
+        var path = Paths.get(LOCAL_CONFIG_DIR);
+        WatchService watchService;
+        try {
+            watchService = initializeWatchService(path);
+            var previousServiceAccounts = readServiceAccounts();
+            WatchKey key;
+            LOGGER.info("wait for changes under {}", LOCAL_CONFIG_DIR);
+            while ((key = watchService.take()) != null) { // take does not return null
+                if (!key.isValid()) {
+                    LOGGER.info("watch key was cancelled or watch service was closed");
+                    LOGGER.info("try to reregister watch");
+                    watchService = initializeWatchService(path);
+                    continue;
+                }
+
+                List<WatchEvent<?>> watchEvents = key.pollEvents();
+                boolean serviceAccountRegistryChanged = false;
+                for (WatchEvent<?> event : watchEvents) {
+                    if (event.context().equals(LOCAL_CONFIG_FILE_NAME)) {
+                        serviceAccountRegistryChanged = true;
+                        break;
                     }
                 }
-            } catch (IOException e) {
-                throw new RegistryException("Failed to addDoguChangeListener", e);
-            } catch (InterruptedException e) {
-                Thread.currentThread().interrupt();
-                throw new RegistryException("Failed to addDoguChangeListener", e);
+
+                if (!serviceAccountRegistryChanged) {
+                    continue;
```

Comment:
If I understand the comment below correctly we need to reset the key in the `continue`-cases too, don't we?

Reply:
Nice catch :+1: